controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java

```diff
@@ -37,16 +72,245 @@ public String getName() {
+        return Futures.toVoid(store.createZNodeIfNotExist(path));
+    }
+
+    CompletableFuture<Void> removeStreamFromScope(String name, int streamPosition) {
+        String path = getPathForStreamPosition(name, streamPosition);
+
+        return Futures.toVoid(store.deletePath(path, true));
+    }
+
+    private String getPathForStreamPosition(String name, int streamPosition) {
+        Preconditions.checkArgument(streamPosition >= 0);
+        Token token = new Token(streamPosition);
+
+        // root path represents most significant 2 digits of the position.
+        String root = ZKPaths.makePath(streamsInScopePath, token.getMsb().toString());
+        // container path represents most significant 3rd to 6th digits of the position.
+        String container = ZKPaths.makePath(root, token.getMiddle().toString());
+        // the path is composed of stream name followed by last 4 chars that represent least significant 4 digits of the position.
+        return ZKPaths.makePath(container, String.format("%s%04d", name, token.getLsb()));
+    }
+
+    private String getStreamName(String x) {
+        return x.substring(0, x.length() - 4);
+    }
+
+    private int getPosition(String x) {
+        return Integer.parseInt(x.substring(x.length() - 4, x.length()));
     }
 
     @Override
     public CompletableFuture<List<String>> listStreamsInScope() {
         return store.getChildren(scopePath);
     }
 
+    @Override
+    public CompletableFuture<Pair<List<String>, String>> listStreams(int limit, String continuationToken, Executor executor) {
+        // Stream references are stored under a hierarchy of nodes as described in `addStreamsInScope method. 
+        // A continuation token is essentially a serialized integer that is broken into three parts - 
+        // msb 2 bytes, middle 4 bytes and lsb 4 bytes. 
+        // stream references are stored as /01/2345/stream-6789. 
+        // So effectively all streams under the scope are ordered by the stream position. Stream position is monotonically 
+        // increasing and any new stream that is added to the scope will be done at a higher position. 
+        // Streams can be deleted though. 
+        // So now the continuation token basically signifies the position of last stream that was returned in previous iteration. 
+        // And continuing from the continuation token, we simply retrieve next `limit` number of streams under the scope. 
+        // Since the streams are stored under the aforesaid hierarchy, we start with all children of the `middle` part of 
+        // continuation token and only include streams whose position is greater than token.
+        // If we are not able to get `limit` number of streams from this, we go to the next higher available `middle` znode 
+        // and fetch its children. If middle is exhausted, we increment the `msb` and repeat until we have found `limit` 
+        // number of streams or reached the end. 
```

Comment:
Thanks for the description, it makes it very clear. I'm now wondering if it would simplify significantly the logic changing the way we determine what stream names to return. What if instead of making it based on a `limit` number of streams, we use the second level of the hierarchy? For example:

- We first return the streams under `00/0000` and continuation token is `00/0001`.
- We second return the streams under `00/0001` and continuation token is `00/0002`
- And so on

Reply:
Streams can be deleted too.. So some of these nodes may only have a very small number of streams. 
And returning from multiple nodes in one query is fine because a) its not complex b) new streams can only be added in a strictly increasing fashion